File: app/src/main/java/com/gmail/cristiandeives/motofretado/ViewMapMvp.java
package com.gmail.cristiandeives.motofretado;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.annotation.StringRes;
import android.support.annotation.UiThread;

import com.gmail.cristiandeives.motofretado.http.Bus;
import com.gmail.cristiandeives.motofretado.http.ModelListener;

import java.util.List;

interface ViewMapMvp {
    interface Model {
        @UiThread
        void readAllBuses(@NonNull ModelListener<Bus[]> listener);
        @UiThread
        void readBus(@NonNull String busId, @NonNull ModelListener<Bus> listener);
        @UiThread
        void cancelAllRequests();
    }

    interface View {
        @UiThread
        String getBusId();
        @UiThread
        void setMapMarker(String title, double latitude, double longitude);
        @UiThread
        void enableBusIdInput();
        @UiThread
        void disableBusIdInput();
        @UiThread
        void displayMessage(@StringRes int messageId);
        @UiThread
        void setAvailableBuses(@NonNull List<Bus> buses, @Nullable String defaultBusId);
    }

    interface Presenter {
        @UiThread
        void onAttach(View view);
        @UiThread
        void onDetach();
        @UiThread
        void startViewingBusLocation();
        @UiThread
        void stopViewingBusLocation();
    }
}

File: app/src/main/java/com/gmail/cristiandeives/motofretado/ViewMapMvp.kt
package com.gmail.cristiandeives.motofretado

import android.support.annotation.StringRes
import android.support.annotation.UiThread
import com.gmail.cristiandeives.motofretado.http.Bus
import com.gmail.cristiandeives.motofretado.http.ModelListener
import java.util.Date

@UiThread
internal interface ViewMapMvp {
    interface Model {
        fun readAllBuses(listener: ModelListener<List<Bus>>)
        fun readBus(busId: String, listener: ModelListener<Bus>)
        fun cancelAllRequests()
    }

    interface View {
        fun getBusId(): String?
        fun setMapMarker(busId: String, latitude: Double, longitude: Double, timestamp: Date?)
        fun enableBusIdInput()
        fun disableBusIdInput()
        fun displayMessage(@StringRes messageId: Int)
        fun setAvailableBuses(buses: List<Bus>, defaultBusId: String?)
        fun setBusError(errorMessage: String)
    }

    interface Presenter {
        fun onAttach(view: View)
        fun onDetach()
        fun startViewingBusLocation()
        fun stopViewingBusLocation()
    }
}